lib/index.js

```diff
@@ -16,204 +16,175 @@ internals.defaults = {
-     - cannot contain space, "*<>:|?
-     - limited to 64 bytes (after conversion to UTF-8)
-     - admin, local and config are reserved
-     */
+        this.settings = this.getSettings(options);
 
-    Hoek.assert(options.partition !== 'admin' && options.partition !== 'local' && options.partition !== 'config', 'Cache partition name cannot be "admin", "local", or "config" when using MongoDB');
-    Hoek.assert(options.partition.length < 64, 'Cache partition must be less than 64 bytes when using MongoDB');
+        this.db = null;
+        this.isConnectionStarted = false;
+        this.isConnected = false;
+        this.collections = {};
+        this.startPending = null;           // Set to an array of callbacks if start pending
+        return this;
+    }
 
-    // merge with defaults
-    const settings = Hoek.applyToDefaults(internals.defaults, options);
+    getSettings(options) {
+        /*
+        Database names:
 
-    // replace or add database
-    settings.uri = settings.uri.replace(/(mongodb:\/\/[^\/]*)([^\?]*)(.*)/,`$1/${settings.partition}$3`);
+        - empty string is not valid
+        - cannot contain space, "*<>:|?
+        - limited to 64 bytes (after conversion to UTF-8)
+        - admin, local and config are reserved
+        */
 
-    return settings;
-};
+        Hoek.assert(options.partition !== 'admin' && options.partition !== 'local' && options.partition !== 'config', 'Cache partition name cannot be "admin", "local", or "config" when using MongoDB');
+        Hoek.assert(options.partition.length < 64, 'Cache partition must be less than 64 bytes when using MongoDB');
 
-internals.Connection.prototype.start = function (callback) {
+        // merge with defaults
+        const settings = Hoek.applyToDefaults(internals.defaults, options);
 
-    // Check if already connected
+        // replace or add database
+        settings.uri = settings.uri.replace(/(mongodb:\/\/[^\/]*)([^\?]*)(.*)/,`$1/${settings.partition}$3`);
 
-    if (this.isConnected) {
-        return callback();
+        return settings;
     }
 
-    // Check if start already pending
-
-    if (this.startPending) {
-        this.startPending.push(callback);
-        return;
-    }
+    async start() {
 
-    // Set start pending state
-
-    this.startPending = [callback];
-
-    const connected = (err) => {
-
-        this.isConnected = !err;
-
-        for (let i = 0; i < this.startPending.length; ++i) {
-            this.startPending[i](err);
+        // Check if already connected
+        if (this.isConnected) {
+            return;
         }
 
-        this.startPending = null;
-    };
-
-    // Connection started flag
-    this.isConnectionStarted = true;
-
-    // Open connection
-
-    MongoDB.MongoClient.connect(this.settings.uri, (err, db) => {
-
-        if (err) {
-            return connected(new Error('Failed opening connection:' + JSON.stringify(err)));
+        // Check if start already pending
+        if (this.startPending) {
```

Comment:
In original implementation it was possible to call `start` multiple times and callback was called for all of them once connected. But this looks like it will just return promise immediately when start is called second time instead of wait for actual connection.

@hueniverse Do I assume correctly that the catbox clients should still handle calling `start` multiple times, which all should resolve in same promise once actual connection promise get resolved?

Reply:
Yes. The cache must be full connected when `start()` resolves no matter how many times it is called. One easy way to do this is to hold on to the promise and return it directly if one has been saved. When you stop the client, you clear out this promise.